src/nsolid/nsolid_heap_snapshot.cc

```diff
@@ -350,5 +402,169 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
+
+  // Construct callFrame
+  json callFrame = {
+    {"functionName", *v8::String::Utf8Value(isolate, node->name)},
+    {"scriptId", node->script_id},
+    {"url", *v8::String::Utf8Value(isolate, node->script_name)},
+    {"lineNumber", node->line_number - 1},
+    {"columnNumber", node->column_number - 1}
+  };
+
+  // Calculate selfSize
+  size_t selfSize = 0;
+  for (const auto& allocation : node->allocations)
+    selfSize += allocation.size * allocation.count;
+
+  // Add properties to result
+  result["callFrame"] = callFrame;
+  result["selfSize"] = selfSize;
+  result["id"] = node->node_id;
+
+  // Add children if there are any
+  if (node->children.empty())
+    return result;
+
+  // Recursively build children
+  nlohmann::ordered_json children;
+  for (const auto* child : node->children)
+    children.push_back(build_sampling_heap_profile_node(isolate, child));
+  result["children"] = children;
+
+  return result;
+}
+
+nlohmann::ordered_json build_v8_allocation_profile(
+    v8::Isolate* isolate,
+    v8::AllocationProfile* profile) {
+  // We need to parse the profile and convert it to a JSON string with the next
+  // schema, this does not support Serializing the profile to a string.
+  // { head: SamplingHeapProfileNode, samples: [SamplingHeapProfileSample] }
+  nlohmann::ordered_json json_profile;
+
+  v8::AllocationProfile::Node* root = profile->GetRootNode();
+  nlohmann::ordered_json head = build_sampling_heap_profile_node(isolate, root);
+  json_profile["head"] = head;
+
+  const std::vector<v8::AllocationProfile::Sample>& samples =
+    profile->GetSamples();
+
+  if (samples.empty()) {
+    json_profile["samples"] = nlohmann::ordered_json::array();
+    return json_profile;
+  }
+
+  for (const auto& sample : samples) {
+    nlohmann::ordered_json sample_json;
+    sample_json["size"] = sample.size * sample.count;
+    sample_json["nodeId"] = sample.node_id;
+    sample_json["ordinal"] = sample.sample_id;
+    json_profile["samples"].push_back(sample_json);
+  }
+
+  return json_profile;
+}
+}  // namespace
+
+
+void NSolidHeapSnapshot::take_sampled_snapshot(
+    SharedEnvInst envinst,
+    NSolidHeapSnapshot* snapshotter) {
+  v8::Isolate* isolate = envinst->isolate();
+  v8::HeapProfiler* profiler = isolate->GetHeapProfiler();
+  ASSERT_NOT_NULL(profiler);
+
+  v8::HandleScope scope(isolate);
+
+  // Get the sampled heap profile
+  v8::AllocationProfile* sampled = profiler->GetAllocationProfile();
+
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  ASSERT(it != snapshotter->threads_running_heap_sampling_.end());
+
+  HeapSamplerStor& stor = it->second;
+  if (sampled == nullptr) {
+    QueueCallback(snapshot_cb,
+                  thread_id,
+                  // TODO(juan) - this is not a heap snapshot failure
+                  heap_profiler::HEAP_SNAPSHOT_FAILURE,
```

Comment:
Should we address this issue before merging?

Reply:
This is just a naming convention that I don't considerate appropriate. That's it.